File: ml-worker/src/monkey_kernel/foresight.py
"""foresight.py — P8 Foresight trajectory predictor.

Stores a rolling trajectory of recent basins per-instance, predicts
the next-step basin via Fisher-Rao geodesic extrapolation, and reports
a regime-adaptive weight per Canonical Principles v2.1 P8.

Pure Fisher-Rao: prediction is slerp(b[-2], b[-1], t=2.0) — one
geodesic step beyond the most recent move. No cosine, no dot product,
no Euclidean distance. Renormalisation is the standard simplex
projection performed by slerp_sqrt itself.

Confidence: smoothness of the recent trajectory measured as
1 / (1 + std(consecutive_distances)). High when basin moves at a
steady pace; low when distances jitter.

Regime weight per P8:
  phi < 0.3                    → linear regime          → 0.1
  equilibrium > 0.7 AND phi<0.3 → breakdown signature   → 0.2
  phi ≥ 0.3                    → geometric regime       → 0.7 × confidence

Pure observation. The trajectory is appended at end of tick (after
basin update); `predict()` is read-only and side-effect free. Tier 6
(Φ-gate selection) is what decides whether to ROUTE on the prediction.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
from statistics import stdev
from typing import Deque, Optional, Tuple

import numpy as np

from qig_core_local.geometry.fisher_rao import fisher_rao_distance, slerp_sqrt

from .persistence import PersistentMemory
from .state import BASIN_DIM


@dataclass(frozen=True)
class ForesightResult:
    """One prediction. predicted_basin is simplex-valid (sum=1, ≥0)
    or zeros when the trajectory is too short to predict.

    Fields:
      predicted_basin : np.ndarray  shape (BASIN_DIM,) on Δ⁶³
      confidence      : float       [0, 1] — trajectory smoothness
      weight          : float       [0, 1] — regime-adaptive blend
      horizon_ms      : float       median tick interval in ms
    """

    predicted_basin: np.ndarray
    confidence: float
    weight: float
    horizon_ms: float


def _empty_result() -> ForesightResult:
    """Cold-start / insufficient-history return: weight=0 means
    callers ignore the prediction."""
    return ForesightResult(
        predicted_basin=np.zeros(BASIN_DIM, dtype=np.float64),
        confidence=0.0,
        weight=0.0,
        horizon_ms=0.0,
    )


def _regime_weight(
    phi: float,
    confidence: float,
    regime_weights: dict[str, float],
) -> float:
    """Per P8: linear (phi<0.3), geometric (phi≥0.3, weight=0.7×conf),
    breakdown (equilibrium>0.7 with low phi → 0.2). Breakdown takes
    precedence over the linear regime since equilibrium-with-low-phi
    is a specific physical signature, not generic uncertainty."""
    eq = regime_weights.get("equilibrium", 0.0)
    if eq > 0.7 and phi < 0.3:
        return 0.2  # breakdown
    if phi < 0.3:
        return 0.1  # linear
    return 0.7 * confidence  # geometric


class ForesightPredictor:
    """One trajectory per instance. Caller owns lifecycle (typically
    one instance per (symbol, lane) tuple in the orchestrator).

    Append is O(1); predict is O(N) where N = trajectory length.
    """

    def __init__(
        self,
        max_trajectory: int = 32,
        *,
        persistence: Optional[PersistentMemory] = None,
        symbol: Optional[str] = None,
    ) -> None:
        self._persistence = persistence
        self._symbol = symbol
        self._traj: Deque[Tuple[np.ndarray, float, float]] = deque(
            maxlen=max_trajectory,
        )
        # Restore prior trajectory from Redis if available.
        if persistence is not None and persistence.is_available and symbol:
            for basin, phi, t_ms in persistence.load_foresight_trajectory(symbol):
                self._traj.append((basin, phi, t_ms))

    def append(self, basin: np.ndarray, phi: float, t_ms: float) -> None:
        """Record a tick. Caller passes the basin AFTER tick update,
        the live phi, and a wall-clock timestamp in ms."""
        b = np.asarray(basin, dtype=np.float64)
        self._traj.append((b, float(phi), float(t_ms)))
        # Write-through.
        if self._persistence is not None and self._symbol:
            self._persistence.push_foresight_step(
                self._symbol, b, float(phi), float(t_ms),
            )

    def predict(self, regime_weights: dict[str, float]) -> ForesightResult:
        """Geodesic extrapolation one tick ahead.

        Trajectory of length 0/1/2 returns empty result (weight=0)
        because the algorithm needs at least 3 points to gauge
        smoothness from pairwise distances.
        """
        if len(self._traj) < 3:
            return _empty_result()

        # Trajectory snapshot for read-only computation
        basins = [item[0] for item in self._traj]
        phis = [item[1] for item in self._traj]
        ts = [item[2] for item in self._traj]

        # Smoothness from consecutive Fisher-Rao distances
        distances = [
            fisher_rao_distance(basins[i], basins[i + 1])
            for i in range(len(basins) - 1)
        ]
        # stdev requires ≥ 2 samples; we guaranteed len(traj)≥3 → len(distances)≥2
        d_std = stdev(distances) if len(distances) >= 2 else 0.0
        confidence = 1.0 / (1.0 + d_std)

        # Geodesic extrapolation: walk one more step beyond b[-1]
        # along the b[-2] → b[-1] direction. slerp_sqrt(p, q, t=2.0)
        # walks twice the p→q distance, projected back onto Δ⁶³.
        predicted = slerp_sqrt(basins[-2], basins[-1], 2.0)

        # Median time delta as horizon estimate
        deltas = [ts[i + 1] - ts[i] for i in range(len(ts) - 1)]
        deltas.sort()
        n = len(deltas)
        horizon_ms = (
            deltas[n // 2]
            if n % 2 == 1
            else 0.5 * (deltas[n // 2 - 1] + deltas[n // 2])
        )

        # Regime-adaptive weight from current phi + regime
        weight = _regime_weight(phis[-1], confidence, regime_weights)

        return ForesightResult(
            predicted_basin=predicted,
            confidence=confidence,
            weight=weight,
            horizon_ms=horizon_ms,
        )

    def reset(self) -> None:
        """Clear the trajectory. Used on regime breaks or test fixtures."""
        self._traj.clear()

    @property
    def trajectory_length(self) -> int:
        return len(self._traj)

File: apps/api/src/services/monkey/foresight.ts
/**
 * foresight.ts — P8 Foresight trajectory predictor (TS parity).
 *
 * Mirrors ml-worker/src/monkey_kernel/foresight.py. One trajectory
 * per instance, geodesic extrapolation via slerp(b[-2], b[-1], 2),
 * regime-adaptive weight per Canonical Principles v2.1 P8.
 *
 * Pure Fisher-Rao: the prediction step uses slerp from basin.ts —
 * no cosine, no dot product, no Euclidean. Renormalisation is
 * standard simplex projection performed inside slerp.
 *
 * Confidence: 1 / (1 + std(consecutive_distances)). High when
 * the basin moves at a steady pace; low when distances jitter.
 *
 * Regime weight per P8:
 *   phi < 0.3                       → linear regime    → 0.1
 *   equilibrium > 0.7 AND phi < 0.3 → breakdown        → 0.2
 *   phi ≥ 0.3                       → geometric        → 0.7 × confidence
 */

import {
  BASIN_DIM,
  fisherRao,
  slerp,
  type Basin,
} from './basin.js';

export interface ForesightResult {
  /** Next-step predicted basin on Δ⁶³, or zero-vector when too cold to predict. */
  predictedBasin: Basin;
  /** [0, 1] — trajectory smoothness. */
  confidence: number;
  /** [0, 1] — regime-adaptive blend weight. */
  weight: number;
  /** Median tick interval in ms (0 when not enough data). */
  horizonMs: number;
}

interface TrajectoryEntry {
  basin: Basin;
  phi: number;
  tMs: number;
}

const emptyResult = (): ForesightResult => ({
  predictedBasin: new Float64Array(BASIN_DIM),
  confidence: 0,
  weight: 0,
  horizonMs: 0,
});

const regimeWeight = (
  phi: number,
  confidence: number,
  regimeWeights: Record<string, number>,
): number => {
  const eq = regimeWeights.equilibrium ?? 0;
  if (eq > 0.7 && phi < 0.3) return 0.2; // breakdown precedence
  if (phi < 0.3) return 0.1; // linear
  return 0.7 * confidence; // geometric
};

const stdev = (xs: number[]): number => {
  const n = xs.length;
  if (n < 2) return 0;
  const mean = xs.reduce((a, b) => a + b, 0) / n;
  // Sample stdev (n-1) to mirror Python's statistics.stdev exactly.
  const v = xs.reduce((a, b) => a + (b - mean) ** 2, 0) / (n - 1);
  return Math.sqrt(v);
};

const median = (xs: number[]): number => {
  if (xs.length === 0) return 0;
  const sorted = [...xs].sort((a, b) => a - b);
  const n = sorted.length;
  return n % 2 === 1
    ? sorted[(n - 1) / 2]
    : 0.5 * (sorted[n / 2 - 1] + sorted[n / 2]);
};

export class ForesightPredictor {
  private readonly trajectory: TrajectoryEntry[] = [];

  constructor(private readonly maxTrajectory: number = 32) {}

  /** Record a tick. Caller passes the basin AFTER tick update. */
  append(basin: Basin, phi: number, tMs: number): void {
    this.trajectory.push({ basin: new Float64Array(basin), phi, tMs });
    while (this.trajectory.length > this.maxTrajectory) {
      this.trajectory.shift();
    }
  }

  /** Geodesic extrapolation one tick ahead. Returns weight=0 result
   * when the trajectory is shorter than 3 entries. */
  predict(regimeWeights: Record<string, number>): ForesightResult {
    if (this.trajectory.length < 3) return emptyResult();

    const basins = this.trajectory.map((e) => e.basin);
    const phis = this.trajectory.map((e) => e.phi);
    const ts = this.trajectory.map((e) => e.tMs);

    // Pairwise Fisher-Rao distances along the trajectory
    const distances: number[] = [];
    for (let i = 0; i < basins.length - 1; i++) {
      distances.push(fisherRao(basins[i], basins[i + 1]));
    }

    const dStd = stdev(distances);
    const confidence = 1 / (1 + dStd);

    // Geodesic extrapolation: walk one step beyond b[-1] along b[-2]→b[-1].
    // slerp(p, q, t=2.0) extends the geodesic, with simplex projection inside.
    const last = basins[basins.length - 1];
    const prev = basins[basins.length - 2];
    const predictedBasin = slerp(prev, last, 2.0);

    // Median tick interval
    const deltas: number[] = [];
    for (let i = 0; i < ts.length - 1; i++) deltas.push(ts[i + 1] - ts[i]);
    const horizonMs = median(deltas);

    const weight = regimeWeight(phis[phis.length - 1], confidence, regimeWeights);

    return { predictedBasin, confidence, weight, horizonMs };
  }

  reset(): void {
    this.trajectory.length = 0;
  }

  get trajectoryLength(): number {
    return this.trajectory.length;
  }
}
